talk/owt/sdk/p2p/p2pclient.cc

```diff
@@ -251,79 +256,88 @@ void P2PClient::OnSignalingMessage(const std::string& message,
           << "Chat cannot be setup since the remote user is not allowed.";
       return;
     }
+    Json::Reader reader;
+    Json::Value json_message;
+    if (!reader.parse(message, json_message)) {
+      RTC_LOG(LS_WARNING) << "Cannot parse incoming message.";
+      return;
+    }
+    std::string message_type;
+    rtc::GetStringFromJsonObject(json_message, kMessageTypeKey, &message_type);
     if (!that->IsPeerConnectionChannelCreated(remote_id)) {
-      if (message.find("\"type\":\"chat-closed\"") != std::string::npos) {
+      if (message_type == "chat-closed") {
         RTC_LOG(LS_WARNING) << "Non-existed chat cannot be stopped.";
         return;
       }
-    } else if (message.find("\"type\":\"offer\"") != std::string::npos) {
-      RTC_LOG(LS_ERROR) << "Received offer from remote.";
-      // If we don't have a PC before we receive an offer.
-      auto pcc = that->GetPeerConnectionChannel(remote_id);
-      if (pcc->HaveLocalOffer() && that->local_id_.compare(remote_id) > 0) {
-        // If our ID is larger than remote, make the remote side as the
-        // publisher (offerer) In case we already have an offer. So we
-        // remove current PCC and create answer.
-        std::shared_ptr<LocalStream> stream = pcc->GetLatestLocalStream();
-        std::function<void()> success_callback =
-            pcc->GetLatestPublishSuccessCallback();
-        std::function<void(std::unique_ptr<Exception>)> failure_callback =
-            pcc->GetLatestPublishFailureCallback();
-        pcc->Stop(nullptr, nullptr);
-        {
-          // If we already created offer,
-          const std::lock_guard<std::mutex> lock(that->pc_channels_mutex_);
-          that->pc_channels_.erase(remote_id);
-        }
-        auto new_pcc = that->GetPeerConnectionChannel(remote_id);
-        new_pcc->OnIncomingSignalingMessage(message);
-        new_pcc->Publish(stream, success_callback, failure_callback);
-        return;
-      }
-    } else if (message.find("\"type\":\"chat-closed\"") != std::string::npos) {
-      RTC_LOG(LS_ERROR) << "Handle the situation that PCC is created by "
-                           "received Chat-closed.";
-      int code = 0;
-      std::string error = "";
-      Json::Reader reader;
-      Json::Value json_message;
-      if (reader.parse(message, json_message)) {
-        Json::Value stop_info;
-        rtc::GetValueFromJsonObject(json_message, "data", &stop_info);
-        rtc::GetIntFromJsonObject(stop_info, "code", &code);
-        rtc::GetStringFromJsonObject(stop_info, "message", &error);
-
-        if (code == kWebrtcIceGatheringPolicyUnsupported) {
-          auto pcc = that->GetPeerConnectionChannel(remote_id);
+    } else if (message_type == "chat-signal"){
+      Json::Value signal;
+      rtc::GetValueFromJsonObject(json_message, kMessageDataKey, &signal);
```

Comment:
You should check if rtc::GetVlueFromJsonObject() returns true before using the signal object. Same for sdp_type parsing.

Reply:
Same as line 266. It briefly scans the message here to handle some special cases. We only care about messages have "data" filed. Other messages will be dispatched to pcc. pcc is the place for processing signaling messages.